File: echo-client.c
#include <sys/types.h>          /* See NOTES */
#include <sys/socket.h>
#include <sys/un.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>

int main() {
  int sockfd = socket(AF_UNIX, SOCK_STREAM, 0);
  struct sockaddr_un addr;
  memset(&addr, 0, sizeof (addr));
  addr.sun_family = AF_UNIX;
  strncpy(addr.sun_path, "/tmp/server", sizeof (addr.sun_path)-1);

  if (connect(sockfd, (struct sockaddr*) &addr, sizeof (addr)) != 0) {
    perror("connect");
    exit(1);
  }

  write(sockfd, "Hello\n", 7);
  char buffer[10];
  read(sockfd, buffer, 10);
  printf("resp: %s", buffer);
}

File: echo-client.zig
// zig build-exe -lc echo-client.zig
const std = @import("std");
const socket = @cImport({
  @cInclude("sys/types.h>");
  @cInclude("sys/socket.h>");
  @cInclude("sys/un.h>");
  @cInclude("unistd.h>");
  @cInclude("stdio.h>");
});
const stdio = @cImport({
  @cInclude("stdio.h");
});

const socketfile = "/tmp/echo-server";

pub fn main() !void {
  const sockfd = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM, 0);
  var addr = std.mem.zeroes(socket.sockaddr_un);
  addr.sun_family = socket.AF_UNIX;
  std.mem.copy(u8, addr.sun_path[0..], socketfile);

  if (socket.connect(sockfd, @ptrCast([*c] const socket.sockaddr, &addr), @sizeOf(socket.sockaddr_un)) != 0) {
    stdio.perror("connect");
    std.os.exit(1);
  }

  _ = socket.write(sockfd, "Hello\n", 7);
  var buffer: [255]u8 = [_]u8{0} ** 255;
  _ = socket.read(sockfd, @ptrCast(*c_void, &buffer), 255);
  std.debug.print("resp: {s}", .{ buffer });
}

